### PR TITLE
[Relay] C++ GraphRuntimeCodegen, Deprecate Python2

### PR DESCRIPTION
**Pull request by antinucleon**
Add C++ GraphRuntimeCodegen.
Deprecate Python GraphRuntimeCodegen.
Deprecate Python2.

### CODE REVIEW

**Comment by antinucleon** [comment]
I am hitting issue with Python2 relative import issue in CI. I am wondering whether we should disable all python2 test according to this RFC(https://github.com/dmlc/tvm/issues/2715), or we should fix this Python2 issue.

@tqchen

File: src/relay/backend/graph_runtime_codegen.cc [hunk 1]
```diff
@@ -0,0 +1,761 @@
...
+    } else {
+      LOG(FATAL) << "type " << checked_type->type_key() << " not supported";
+    }
+    return {GraphNodeRef(node_id, 0)};
+  }
+
+  /*! \brief Visitors */
+  std::unordered_map<Expr, std::vector<GraphNodeRef>, NodeHash, NodeEqual> visitor_cache_;
+
+  std::vector<GraphNodeRef> VisitExpr(const Expr& expr) override {
```
Comment (jroesch) [inline src/relay/backend/graph_runtime_codegen.cc]: Do we need to overload this function?, this should happen by default

File: src/relay/backend/graph_runtime_codegen.cc [hunk 2]
```diff
@@ -0,0 +1,761 @@
...
+  /*! \brief lowered funcs */
+  std::unordered_map<std::string, std::unordered_set<LoweredFunc, NodeHash, NodeEqual>>
+      lowered_funcs_;
+  /*! \brief name map */
+  std::unordered_map<std::string, size_t> name_map_;
+  /*! \brief compile engine */
+  CompileEngine compile_engine_;
+};
+
+class GraphRuntimeCodegenModule : public runtime::ModuleNode {
```
Comment (jroesch) [inline src/relay/backend/graph_runtime_codegen.cc]: I think this makes sense for now, but in general I'm not a huge fan of reusing tvm::Module to implement this stringly typed function lookup map. Is there a better way to do this @tqchen ?

**Comment by jroesch** [comment]
Just some minor comments from me.

File: src/relay/backend/graph_runtime_codegen.cc (hunk 1 above)
Comment (antinucleon) [reply src/relay/backend/graph_runtime_codegen.cc]: I think we need.

File: src/relay/backend/graph_runtime_codegen.cc (hunk 2 above)
Comment (antinucleon) [reply src/relay/backend/graph_runtime_codegen.cc]: It is a stateful function. I think Module is a convenient & clean way to do this.

**Comment by antinucleon** [comment]
@tqchen After switch to Python3, doc build fails for some reason of RPC to localhost. I can't reproduce it locally. On CI both ```localhost``` and ```0.0.0.0``` failed when building doc, but worked well in unittest. Could you help take a look of this issue? It is the only blocker of this PR,

**Comment by antinucleon** [comment]
@jroesch @tqchen I have fixed Python3 doc build issue.

**Comment by antinucleon** [comment]
@jroesch Relay C++ Build module depends on this PR. Please let me know if there is anything more needed for this PR.

File: tests/scripts/task_python_integration.sh [hunk 3]
```diff
@@ -25,22 +25,17 @@ export LD_LIBRARY_PATH="build:${LD_LIBRARY_PATH:-}"
...
 # Test TVM
-make cython
 make cython3
 
 # Test extern package
 cd apps/extension
 rm -rf lib
 make
 cd ../..
-python -m nose -v apps/extension/tests
```
Comment (wweic) [inline tests/scripts/task_python_integration.sh]: should change to `python3` instead of deleting, right?

Comment (antinucleon) [reply tests/scripts/task_python_integration.sh]: Fixed.

File: src/relay/backend/graph_runtime_codegen.cc (hunk 1 above)
Comment (jroesch) [reply src/relay/backend/graph_runtime_codegen.cc]: Oh right caching only exists for the Mutator/Visitor

File: src/relay/backend/graph_runtime_codegen.cc [hunk 4]
```diff
@@ -0,0 +1,761 @@
...
+  }
+  std::vector<GraphNodeRef> VisitExpr_(const CallNode* op) override {
+    Expr expr = GetRef<Expr>(op);
+    Function func;
+    if (op->op.as<OpNode>()) {
+      LOG(FATAL) << "Operators should be transformed away; try applying"
+                 << "the fuse_ops transformation to the expression.";
+    } else if (op->op.as<GlobalVarNode>()) {
+      LOG(FATAL) << "Not implemented";
+      // func = self.mod[call.op]
```
Comment (wweic) [inline src/relay/backend/graph_runtime_codegen.cc]: remove.

File: src/relay/backend/graph_runtime_codegen.cc (hunk 2 above)
Comment (jroesch) [reply src/relay/backend/graph_runtime_codegen.cc]: Yeah I understand, the string based dispatch of module isn't my favorite design we just did the same thing in the VTA simulator stuff.

**Comment by jroesch** [comment]
@antinucleon can you just fix the last comment from @wweic then we can merge?

File: src/relay/backend/graph_runtime_codegen.cc (hunk 4 above)
Comment (antinucleon) [reply src/relay/backend/graph_runtime_codegen.cc]: Updated

**Comment by Laurawly** [comment]
Thanks @antinucleon @jroesch @wweic, this is now merged.

**Pull request merged**